zebra-state/src/service/read/tree.rs

```diff
@@ -164,96 +89,93 @@ where
-/// different fork to `db`, then the trees can be inconsistent. In that case, we ignore all the
-/// trees in `chain` after the first inconsistent tree, because we know they will be inconsistent as
-/// well. (It is cryptographically impossible for tree roots to be equal once the leaves have
-/// diverged.)
-/// 2. APIs that return single subtrees can't be used here, because they can create
-/// an inconsistent list of subtrees after concurrent non-finalized and finalized updates.
+/// If there is no subtree at the first index in the range, the returned list is empty.
+/// Otherwise, subtrees are continuous up to the finalized tip.
 pub fn orchard_subtrees<C>(
     chain: Option<C>,
     db: &ZebraDb,
-    start_index: NoteCommitmentSubtreeIndex,
-    limit: Option<NoteCommitmentSubtreeIndex>,
+    range: impl std::ops::RangeBounds<NoteCommitmentSubtreeIndex> + Clone,
 ) -> BTreeMap<NoteCommitmentSubtreeIndex, NoteCommitmentSubtreeData<orchard::tree::Node>>
 where
     C: AsRef<Chain>,
 {
-    let mut db_list = db.orchard_subtree_list_by_index_for_rpc(start_index, limit);
-
-    if let Some(limit) = limit {
-        let subtrees_num = u16::try_from(db_list.len())
-            .expect("There can't be more than `u16::MAX` Orchard subtrees.");
-
-        // Return the subtrees if the amount of them reached the given limit.
-        if subtrees_num == limit.0 {
-            return db_list;
-        }
-
-        // If not, make sure the amount is below the limit.
-        debug_assert!(subtrees_num < limit.0);
-    }
-
-    // If there's no chain, then we have the complete list.
-    let Some(chain) = chain else {
-        return db_list;
-    };
+    subtrees(
+        chain,
+        range,
+        |chain, range| chain.orchard_subtrees_in_range(range),
+        |range| db.orchard_subtree_list_by_index_range(range),
+    )
+}
 
-    // Unlike the other methods, this returns any trees in the range,
-    // even if there is no tree for start_index.
-    let fork_list = chain.as_ref().orchard_subtrees_in_range(start_index, limit);
+/// Returns a list of [`NoteCommitmentSubtree`]s starting at `start_index`.
+///
+/// If there is no subtree at the first index in the range, the returned list is empty.
+/// Otherwise, subtrees are continuous up to the finalized tip.
+///
+/// # Correctness
+///
+/// APIs that return single subtrees can't be used here, because they can create
+/// an inconsistent list of subtrees after concurrent non-finalized and finalized updates.
+fn subtrees<C, R, N, F1, F2>(
+    chain: Option<C>,
+    range: R,
+    read_chain: F1,
+    read_disk: F2,
+) -> BTreeMap<NoteCommitmentSubtreeIndex, NoteCommitmentSubtreeData<N>>
+where
+    C: AsRef<Chain>,
+    N: PartialEq,
+    R: std::ops::RangeBounds<NoteCommitmentSubtreeIndex> + Clone,
+    F1: FnOnce(&Chain, R) -> BTreeMap<NoteCommitmentSubtreeIndex, NoteCommitmentSubtreeData<N>>,
+    F2: FnOnce(R) -> BTreeMap<NoteCommitmentSubtreeIndex, NoteCommitmentSubtreeData<N>>,
+{
```

Comment:
Optional Nitpick: As a reviewer, I prefer complicated generic functions to have meaningful names for their generic types. Otherwise I spend a lot of time working out what each type does.

For example, it took me more than a minute to work out that `N` was the type of the tree note. This could be avoided by renaming `N` to `Note`, or whatever the generic is called in `NoteCommitmentSubtreeData<_>`.

I'd also suggest renaming:
* F1 to ChainSubtreeFn
* F2 to DbSubtreeFn
* Optional: C to Chain
* Optional: R to Range

Reply:
> As a reviewer, I prefer complicated generic functions to have meaningful names for their generic types. Otherwise I spend a lot of time working out what each type does.

Whoops. Noted.

Renamed in https://github.com/ZcashFoundation/zebra/pull/7739/commits/668462d586cfe5a409c56b189478c374296e6b2d.